simde/x86/sse4.2.h

```diff
@@ -37,6 +37,231 @@ SIMDE__BEGIN_DECLS
 #  define SIMDE_X86_SSE4_2_ENABLE_NATIVE_ALIASES
 #endif
 
+#define SIMDE_SIDD_CMP_EQUAL_ANY     0
+#define SIMDE_SIDD_CMP_RANGES     1
+#define SIMDE_SIDD_CMP_EQUAL_EACH     2
+#define SIMDE_SIDD_CMP_EQUAL_ORDERED   3
+
+#if defined(SIMDE_X86_SSE4_2_ENABLE_NATIVE_ALIASES)
+#define _SIDD_CMP_EQUAL_ANY SIMDE_SIDD_CMP_EQUAL_ANY
+#define _SIDD_CMP_RANGES SIMDE_SIDD_CMP_RANGES
+#define _SIDD_CMP_EQUAL_EACH SIMDE_SIDD_CMP_EQUAL_EACH
+#define _SIDD_CMP_EQUAL_ORDERED SIMDE_SIDD_CMP_EQUAL_ORDERED
+#endif
+
+SIMDE__FUNCTION_ATTRIBUTES
+int
+simde_mm_cmpestra_8_(simde__m128i a, int la, simde__m128i b, int lb, const int imm8) {
+  const int cmp_op = imm8 & 0x06;
+  const int polarity = imm8 & 0x30;
+  simde__m128i_private
+    bool_res_ = simde__m128i_to_private(simde_mm_setzero_si128()),
+    a_ = simde__m128i_to_private(a),
+    b_ = simde__m128i_to_private(b);
+  const int upper_bound = (128 / 8) - 1;
+  int a_invalid = 0;
+  int b_invalid = 0;
+  for(int i = 0 ; i < (upper_bound) ; i++) {
+    for(int j = 0; j< (upper_bound) ; j++){
+      int bitvalue = ((a_.i8[i] == b_.i8[j]) ? 1 : 0);
+      bool_res_.i8[i] |= (( bitvalue ) << j);
+      if(i == la)
+        a_invalid = 1;
+      if(j == lb)
+        b_invalid = 1;
+      switch(cmp_op){
+        case SIMDE_SIDD_CMP_EQUAL_ANY:
+          break;
+        case SIMDE_SIDD_CMP_RANGES:
+          break;
+        case SIMDE_SIDD_CMP_EQUAL_EACH:
+          if(a_invalid && b_invalid)
+            bool_res_.i8[i] |= (1 << j);
+          break;
+        case SIMDE_SIDD_CMP_EQUAL_ORDERED:
+          if(a_invalid && !b_invalid)
+            bool_res_.i8[i] |= (1 << j);
+          else if(a_invalid && b_invalid)
+            bool_res_.i8[i] |= (1 << j);
+          break;
+      }
+    }
+  }
+  int32_t int_res_1 = 0;
+  int32_t int_res_2 = 0;
+  switch(cmp_op) {
+    case SIMDE_SIDD_CMP_EQUAL_ANY:
+      for(int i = 0 ; i < (upper_bound) ; i++){
+        SIMDE__VECTORIZE_REDUCTION(|:int_res_1)
+        for(int j = 0 ; j < (upper_bound) ; j++){
+          int_res_1 |= (((bool_res_.i8[i] >> j) & 1) << i);
+        }
+      }
+      break;
+    case SIMDE_SIDD_CMP_RANGES:
+      for(int i = 0 ; i < (upper_bound) ; i++){
+        SIMDE__VECTORIZE_REDUCTION(|:int_res_1)
+        for(int j = 0 ; j < (upper_bound) ; j++){
+          int_res_1 |= ((((bool_res_.i8[i] >> j) & 1) & ((bool_res_.i8[i] >> (j + 1)) & 1)) << i);
+          j += 2;
+        }
+      }
+      break;
+    case SIMDE_SIDD_CMP_EQUAL_EACH:
+      for(int i = 0 ; i < (upper_bound) ; i++){
+        SIMDE__VECTORIZE_REDUCTION(|:int_res_1)
+        for(int j = 0 ; j < (upper_bound) ; j++){
+          int_res_1 |= (((bool_res_.i8[i] >> i) & 1) << i);
+        }
+      }
+      break;
+    case SIMDE_SIDD_CMP_EQUAL_ORDERED:
+      int_res_1 = (imm8 & 1) ? 0xff : 0xffff;
```

Comment:
We're in the 8-bit function, so we already know what the character size is.

Also, use the macros to keep it easy to read.  We don't want to force people reading the code to remember what each bit means.  Besides, in this case it's actually wrong; 1 means 16-bit characters in the Intel API, and we have to match the Intel API since someone using the API might just do `_mm_cmpestra(..., 1)` instead of using the macros.

Reply:
changed it accordingly,
but what is the reason for it to be different from intrinsics guide ?

Reply:
I don't know why the IIG does what it does, but you should never feel bound by it.  The pseudo-code is only there to explain what the intrinsic does; if we can do something better, faster, or simpler in SIMDe then we should.

In this case, there is no benefit to the extra step; the compiler will optimize it away, but it still requires people reading the code to keep it in their head.